### PR TITLE
Improve `Mesh` documentation

### PR DESCRIPTION
**Pull request by Selene-Amanita**
# Objective

This PR continues https://github.com/bevyengine/bevy/pull/8885

It aims to improve the `Mesh` documentation in the following ways:
- Put everything at the "top level" instead of the "impl".
- Explain better what is a Mesh, how it can be created, and that it can be edited.
- Explain it can be used with a `Material`, and mention `StandardMaterial`, `PbrBundle`, `ColorMaterial`, and `ColorMesh2dBundle` since those cover most cases
- Mention the glTF/Bevy vocabulary discrepancy for "Mesh"
- Add an image for the example
- Various nitpicky modifications

## Note

- The image I added is 90.3ko which I think is small enough?
- Since rustdoc doesn't allow cross-reference not in dependencies of a subcrate [yet](https://github.com/rust-lang/rust/issues/74481), I have a lot of backtick references that are not links :(
- Since rustdoc doesn't allow linking to code in the crate (?) I put link to github directly.
- Since rustdoc doesn't allow embed images in doc [yet](https://github.com/rust-lang/rust/issues/32104), maybe [soon](https://github.com/rust-lang/rfcs/pull/3397), I had to put only a link to the image. I don't think it's worth adding [embed_doc_image](https://docs.rs/embed-doc-image/latest/embed_doc_image/) as a dependency for this.

### CODE REVIEW

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 1]
```diff
@@ -26,28 +26,24 @@ use wgpu::{
...
-    /// which allows easy stable VertexBuffers (i.e. same buffer order)
-    attributes: BTreeMap<MeshVertexAttributeId, MeshAttributeData>,
-    indices: Option<Indices>,
-    morph_targets: Option<Handle<Image>>,
-    morph_target_names: Option<Vec<String>>,
-}
-
-/// Contains geometry in the form of a mesh.
+/// A 3D object made out of vertices representing triangles, lines, or points,
+/// with special "attribute" values for each vertex.
```
Comment (IceSentry) [inline crates/bevy_render/src/mesh/mesh/mod.rs:30]: I'm not sure I'd consider attributes to be special. They are an integral part of meshes.

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 2]
```diff
@@ -78,50 +74,82 @@ pub struct Mesh {
...
+/// used in a `PbrBundle` with a square bevy logo texture, with added axis, points,
+/// lines and text for clarity.
+///
+/// For further visualization, explanation, and examples see the built-in Bevy examples,
+/// in particular [generate_custom_mesh](https://github.com/bevyengine/bevy/blob/main/examples/3d/generate_custom_mesh.rs),
+/// and the [implementation of the built-in shapes](https://github.com/bevyengine/bevy/tree/main/crates/bevy_render/src/mesh/shape).
+///
+/// ## Common points of confusion
 ///
+/// - UV maps in Bevy are "flipped", `[0.0, 0.0]` = Top-Left (not Bot-Left like `OpenGL`).
```
Comment (IceSentry) [inline crates/bevy_render/src/mesh/mesh/mod.rs:90]: It's only flipped compared to OpenGL, to me this implies bevy is doing something non-standard here. As far as I know, top-left is part of the gltf spec

In other words, I agree it's worth pointing out, but it should be clearer that it's flipped compared to opengl.

Something like
```suggestion
/// - UV maps in bevy start at the top-left unlike opengl which starts at the bottom-left
```

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 3]
```diff
@@ -78,50 +74,82 @@ pub struct Mesh {
...
+/// [0.0, 0.0, 1.0] is very common for simple flat meshes on the XY plane,
+/// because simple meshes are smooth and they don't require complex light calculations.
+/// - Vertex winding order: by default, `StandardMaterial.cull_mode` is [`Some(Face::Front)`](crate::render_resource::Face),
+/// which means that Bevy would *only* render the "front" of each triangle, which
+/// is the side of the triangle from where the vertices appear in a *counter-clockwise* order.
+///
+// TODO: allow values to be unloaded after been submitting to the GPU to conserve memory
+#[derive(Debug, TypeUuid, TypePath, Clone)]
+#[uuid = "8ecbac0f-f545-4473-ad43-e1f4243af51e"]
+pub struct Mesh {
```
Comment (IceSentry) [inline crates/bevy_render/src/mesh/mesh/mod.rs:109]: Since you're already improving the docs. It might be worth it to also add short example of actually getting back an attribute from the mesh. The code to do that can be a little convoluted so having a small code block showing it could be useful.

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 4]
```diff
@@ -78,50 +74,82 @@ pub struct Mesh {
...
+/// lines and text for clarity.
+///
+/// For further visualization, explanation, and examples see the built-in Bevy examples,
+/// in particular [generate_custom_mesh](https://github.com/bevyengine/bevy/blob/main/examples/3d/generate_custom_mesh.rs),
+/// and the [implementation of the built-in shapes](https://github.com/bevyengine/bevy/tree/main/crates/bevy_render/src/mesh/shape).
+///
+/// ## Common points of confusion
 ///
+/// - UV maps in Bevy are "flipped", `[0.0, 0.0]` = Top-Left (not Bot-Left like `OpenGL`).
+/// - It is normal for multiple vertices to have the same [position attribute](Mesh::ATTRIBUTE_POSITION) value,
```
Comment (IceSentry) [inline crates/bevy_render/src/mesh/mesh/mod.rs:91]: Maybe try to rephrase this to not use the word normal? I generally wouldn't mind, but in the context of meshes it might be worth avoiding the potential confusing with vertex normal.

**Review by IceSentry — COMMENTED** [review]
Couple of nits. Personally, this feels like it's touching a bit too many things, like I'm not sure the AssetServer or StandardMaterial should really be mentioned here but this really isn't targeted at someone like me so I'm not sure what the best solution is.

File: crates/bevy_render/src/mesh/mesh/mod.rs (hunk 3 above)
Comment (Selene-Amanita) [reply crates/bevy_render/src/mesh/mesh/mod.rs:109]: This is covered in the `generate_custom_mesh` example that I forgot to link in the PR description: https://github.com/bevyengine/bevy/pull/8909

But I did link to it in this PR code, under the doc example.

I don't think it's not worth making a really big doc example, but I can clarify that `generate_custom_mesh` does that!

Comment (IceSentry) [reply crates/bevy_render/src/mesh/mesh/mod.rs:109]: Ah, good to know, I didn't know that was in that example

**Comment by Selene-Amanita** [comment]
Thank you for your review!

I tried to fix stuff from your comments in the last commit.
About the "flipped UV" and use of "normal" those weren't my sentences to begin with but I agree it's clearer that way!

> Couple of nits. Personally, this feels like it's touching a bit too many things, like I'm not sure the AssetServer or StandardMaterial should really be mentioned here but this really isn't targeted at someone like me so I'm not sure what the best solution is.

I think people might want to use a Mesh in Bevy, have no idea how to, seach "Mesh Bevy", and either find this doc, or Bevy's cheatbook, in the first case I think it's good to have some references to them at least.

Also knowing that the "UV" and "Normals" are important to put in a mesh for StandardMaterial to work properly is also important imo.

Tell me if it still needs improvement.

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 5]
```diff
@@ -78,50 +74,87 @@ pub struct Mesh {
...
+///
+/// ## Use with `StandardMaterial`
+///
+/// To render correctly with `StandardMaterial`, a mesh needs to have properly defined:
+/// - [`UVs`](Mesh::ATTRIBUTE_UV_0): Bevy needs to know how to map a texture onto the mesh
+/// (also true for `ColorMaterial`).
+/// - [`Normals`](Mesh::ATTRIBUTE_NORMAL): Bevy needs to know how light interacts with your mesh.
+/// [0.0, 0.0, 1.0] is very common for simple flat meshes on the XY plane,
+/// because simple meshes are smooth and they don't require complex light calculations.
+/// - Vertex winding order: by default, `StandardMaterial.cull_mode` is [`Some(Face::Front)`](crate::render_resource::Face),
```
Comment (konsolas) [inline crates/bevy_render/src/mesh/mesh/mod.rs:107]: This is not correct; see https://github.com/bevyengine/bevy/blob/main/crates/bevy_pbr/src/pbr_material.rs#L191, it should be Face::Back.

The rest of the explanation in this bullet point is still correct, however.

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 6]
```diff
@@ -78,50 +74,87 @@ pub struct Mesh {
...
-///              common for simple meshes because simple meshes are smooth, and they don't require complex light calculations.)
-/// - Vertex winding order -
-///              the default behavior is with `StandardMaterial.cull_mode` = Some([`Face::Front`](crate::render_resource::Face::Front)) which means
-///              that by default Bevy would *only* render the front of each triangle, and the front
-///              is the side of the triangle in which the vertices appear in a *counter-clockwise* order.
+/// You can see how it looks like [here](https://github.com/bevyengine/bevy/blob/main/assets/dovs/Mesh.png),
+/// used in a `PbrBundle` with a square bevy logo texture, with added axis, points,
+/// lines and text for clarity.
+///
+/// ## Other exampless
```
Comment (konsolas) [inline crates/bevy_render/src/mesh/mesh/mod.rs:84]: exampless

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 7]
```diff
@@ -78,50 +74,87 @@ pub struct Mesh {
...
-    /// Use in conjunction with [`Mesh::insert_attribute`]
+    /// Use in conjunction with [`Mesh::insert_attribute`].
     pub const ATTRIBUTE_NORMAL: MeshVertexAttribute =
         MeshVertexAttribute::new("Vertex_Normal", 1, VertexFormat::Float32x3);
 
-    /// Texture coordinates for the vertex. Use in conjunction with [`Mesh::insert_attribute`]
+    /// Texture coordinates for the vertex. Use in conjunction with [`Mesh::insert_attribute`].
+    ///
+    /// Values are generally between 0. and 1., `[0.,0.]` being the top left of the texture,
+    /// and [1.,1.] the bottom-right. Values outside this range will generate visual artifacts with
```
Comment (konsolas) [inline crates/bevy_render/src/mesh/mesh/mod.rs:139]: I believe this should read "Other values will be clamped to this range by default".

**Review by konsolas — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/mesh/mesh/mod.rs (hunk 7 above)
Comment (Selene-Amanita) [reply crates/bevy_render/src/mesh/mesh/mod.rs:139]: Well they're not clamped it makes an artifact. it "expands" the texture with the color value it has at the closest border

If I put "2.0" instead of "1.0" for y UV values for example:
image

I'll fix the typo and mistake you mentioned earlier, thank you!

Comment (konsolas) [reply crates/bevy_render/src/mesh/mesh/mod.rs:139]: Yes, this is the behaviour you would expect if the texture coordinates are clamped to [0, 1]. The value of the texel at the closest edge of the texture will be used.

This behaviour is specified by WebGPU - https://gpuweb.github.io/gpuweb/#enumdef-gpuaddressmode
It just happens to result in that artifact with the texture you chose, but it might produce reasonable results for other cases.

Comment (Selene-Amanita) [reply crates/bevy_render/src/mesh/mesh/mod.rs:139]: Okay but it's clamped by the rendering process, for all the pixels with outside value, not for each vertex, otherwise putting `2.` would have the same effect as putting `1.`.

In general, for the end user:
- this is something they don't want
- it will be perceived as a visual artifact
- if we tell them the value is clamped they might get the wrong impression that it doesn't matter if they put values outside the range

I think going into details is more confusing than just saying "don't do it" here.

Comment (konsolas) [reply crates/bevy_render/src/mesh/mesh/mod.rs:139]: I believe the main use of this behaviour is producing a solid border color. For example, the following logo has a solid red border, and sampling it with uv coordinates outside the 0-1 range can be used to extend this border to cover a surface that is not the same shape without warping the logo:

kfc logo

But I'm happy to defer to other reviewers here if the consensus is that it confuses the end user.

Comment (Selene-Amanita) [reply crates/bevy_render/src/mesh/mesh/mod.rs:139]: Fair enough, I'll try to change the formulation to explain this.

Comment (konsolas) [reply crates/bevy_render/src/mesh/mesh/mod.rs]: looks good to me

**Review by konsolas — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
